statsrelay/src/backend_client/statsd_client.rs

```diff
@@ -168,14 +171,35 @@ async fn client_sender(
         form_connection(stats.clone(), endpoint.as_str(), first_connect_tripwire).await;
 
     loop {
-        // Wait for a channel message ready for network transmission
-        let mut buf = match recv.recv().await {
-            None => {
-                debug!("sender task {} exiting", endpoint);
+        // We have a connection but this could get stale before we receive data.
+        // First check if there's data ready and if not, give time for it to appear
+        // before disconnecting the current connection so the next recv() forces a reconnect.
+        // This shouldn't be necessary but there's an issue with some of the buffer
```

Comment:
I found a [thread](https://users.rust-lang.org/t/tcpstream-write-silently-loses-one-message/38206) describing this exact issue.  This is actually a characteristic of TCP connections.  I think closing the connection or adding keep alive (which seems to not be primitively supported in rust) is the right answer here, not necessarily just a hack 🙂

Reply:
Nice find! I found a similar thread about this issue on the read side but didn't find this one about the write. Clears some things up.